pkg/cache/lrucache.go

```diff
@@ -214,13 +214,23 @@ func (c *GitRepoLRUCache) tryEvict() error {
 			break
 		}
 
+		// Check that the LRU element is not part of neverEvictRepos
+		// If it is, find the nearest LRU not part of neverEvictRepos
+		lru := c.dll.Back()
+		for _, isInNeverEvictRepos := neverEvictRepos[lru.Value.(*GitRepoFilePath).key]; isInNeverEvictRepos; {
+			lru = lru.Next()
+			if lru == nil {
+				return fmt.Errorf("Disk space completely occupied by neverEvictRepos, could not evict")
+			}
+		}
```

Comment:
This is a broad comment on the efficiency of this approach:

This essentially makes eviction a O n*m operation where n is the number of elements in the lru cache and m is the number of operations to perform to clear enough disk space to meet the `minFreeBytes`. Given a worst case scenario, where disk space has filled up and needs many m operations to clear it, and we must iterate many n number of elements in the lru cache to find one that's not in the `neverEvictRepos`.

That's probably fine but maybe in the future we'll look at optimizing this.

Reply:
Correct, this will be O(n*m) always in the worst case. I think it can be optimized to O(mlogn) for worst case, if we cache the node that comes after the first LRU is found, then we "half" the space for the next search, and it might be possible since the cache gets locked for the eviction. The ordering of the cache does not get shuffled while the eviction algo is happening due to it being concurrent safe, right?

Reply:
>  The ordering of the cache does not get shuffled while the eviction algo is happening due to it being concurrent safe, right?

Good question: as the server runs and the cache fills up, i believe with the way you have it implemented, the _front_ of the cache will be all the repos that shouldn't be evicted and the back will have the new repos. So after running for a long time, the performance of this chunk will get worse.

Which is probably fine. There's lots of ways for users to fine tune this: i.e., not have as many repos in the never evict list.